component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -0,0 +1,222 @@
+                return false;
+            }
+            return true;
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+    }
+
+    /**
+     * Returns whether the incoming request can be handled by the particular authenticator.
+     *
+     * @param httpServletRequest      http servelet request
+     * @param content                 map of request body params
+     * @param oAuthClientAuthnContext oAuthClientAuthnContext
+     * @return true if the incoming request can be handled.
+     */
+    @Override
+    public boolean canAuthenticate(HttpServletRequest httpServletRequest, Map<String, List> content,
+                                   OAuthClientAuthnContext oAuthClientAuthnContext) {
+
+        String clientAssertionType = null;
+        String clientAssertion = null;
+        List oauthJWTAssertionType = content.get(OAUTH_JWT_ASSERTION_TYPE);
+        List oauthJWTAssertion = content.get(OAUTH_JWT_ASSERTION_TYPE);
+        if (oauthJWTAssertionType != null && oauthJWTAssertionType.get(0) != null) {
+            clientAssertionType = oauthJWTAssertionType.get(0).toString();
+        }
+        if (oauthJWTAssertion != null && oauthJWTAssertion.get(0) != null) {
+            clientAssertion = oauthJWTAssertion.get(0).toString();
+        }
+
+        return isValidJWTClientAssertionRequest(clientAssertionType, clientAssertion);
+    }
+
+    /**
+     * Retrievs the client ID which is extracted from the JWT.
+     *
+     * @param httpServletRequest
+     * @param content
+     * @param oAuthClientAuthnContext
+     * @return jwt 'sub' value as the client id
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public String getClientId(HttpServletRequest httpServletRequest, Map<String, List> content,
+                              OAuthClientAuthnContext oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        SignedJWT signedJWT = null;
+        try {
+            List oauthJWTAssertion = content.get(OAUTH_JWT_ASSERTION);
+            if (oauthJWTAssertion != null && oauthJWTAssertion != null) {
+                signedJWT = getSignedJWT(oauthJWTAssertion.toString());
+            }
+            ReadOnlyJWTClaimsSet claimsSet = jwtValidator.getClaimSet(signedJWT);
+            String jwtSubject = jwtValidator.resolveSubject(claimsSet);
+            if (jwtValidator.validateSubject(jwtSubject)) {
+                oAuthClientAuthnContext.setClientId(jwtSubject);
+            }
+
+        } catch (IdentityOAuth2Exception e) {
+            String message = "Error while obtaining the client id from the token request.";
+            throw new OAuthClientAuthnException(message, OAuth2ErrorCodes.INVALID_CLIENT);
+        }
+        return oAuthClientAuthnContext.getClientId();
+    }
+
+    private SignedJWT getSignedJWT(String assertion)
+            throws IdentityOAuth2Exception {
```

Comment:
Can we bring this line to the previous line ? I don't think it will go beyond limit 120